### PR TITLE
[REEF-1875] Upgrade Netty depenency to 4.1.15Final

### PR DESCRIPTION
**Pull request by nhne**
* removed a deprecated method `ByteBuf.order`
* changed name of exception caught in `open` in `NettyMessagingTransport`.

JIRA:
[REEF-1875](https://issues.apache.org/jira/browse/REEF-1875)

Pull request:
This closes #1371

### CODE REVIEW

**Comment by bgchun** [comment]
@nhne LGTM. I will test and merge this PR.

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/ChunkedReadWriteHandler.java [hunk 1]
```diff
@@ -147,9 +147,9 @@ public void write(final ChannelHandlerContext ctx, final Object msg, final Chann
    */
   private byte[] sizeAsByteArr(final int size) {
     final byte[] ret = new byte[INT_SIZE];
-    final ByteBuf intBuffer = Unpooled.wrappedBuffer(ret).order(Unpooled.LITTLE_ENDIAN);
+    final ByteBuf intBuffer = Unpooled.wrappedBuffer(ret);
```
Comment (bgchun) [inline lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/ChunkedReadWriteHandler.java]: Any particular reason not to use `order`?

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/ChunkedReadWriteHandler.java [hunk 2]
```diff
@@ -170,8 +170,8 @@ private int getSize(final byte[] data, final int offset) {
       return 0;
     }
 
-    final ByteBuf intBuffer = Unpooled.wrappedBuffer(data, offset, INT_SIZE).order(Unpooled.LITTLE_ENDIAN);
-    final int ret = intBuffer.readInt();
+    final ByteBuf intBuffer = Unpooled.wrappedBuffer(data, offset, INT_SIZE);
```
Comment (bgchun) [inline lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/ChunkedReadWriteHandler.java]: Any particular reason not to use `order`?

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/ChunkedReadWriteHandler.java (hunk 1 above)
Comment (nhne) [reply lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/ChunkedReadWriteHandler.java]: The complier warned me that `ByteBuf.order` has deprecated. I followed the instruction at [Netty docs](https://netty.io/4.1/api/io/netty/buffer/ByteBuf.html#order-java.nio.ByteOrder-).

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java [hunk 3]
```diff
@@ -320,7 +320,7 @@ public void close() {
         }
         break;
       } catch (final Exception e) {
-        if (e.getClass().getSimpleName().compareTo("ConnectException") == 0) {
+        if (e.getClass().getSimpleName().compareTo("AnnotatedConnectException") == 0) {
```
Comment (bgchun) [inline lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java]: Did you actually simulate the "connection refused" event to check the simple name is `AnnotatedConnectException`? 
Can't we do `e instanceof AnnotatedConnectException`?

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/ChunkedReadWriteHandler.java (hunk 2 above)
Comment (nhne) [reply lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/ChunkedReadWriteHandler.java]: Same as above.

**Comment by bgchun** [comment]
@nhne I left comments. Before merging it, I'd like to check a few things. Please take a look.

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java (hunk 3 above)
Comment (nhne) [reply lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java]: I ran the test and found that the exception's simple name was changed to `connectionException` to `AnnotatedConnecException`. The exception was raised both version of Netty as "connection refused", but exception was changed.

I'll test I can do it as `e instanceof AnnotatedConnectException`. Thanks for your idea!

Comment (bgchun) [reply lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java]: Or you can try the following.
'''
} catch (final AnnotatedConnectException e) {
  ...
} catch (final Exception e) {
  throw e;
}
'''

Comment (nhne) [reply lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java]: Unfortunately, the `AnnotatedConnectException` is private class of `io.netty.channel.AbstractChannel` and cannot be accessed in public. Could you please tell me other ideas for this?

Comment (bgchun) [reply lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java]: Never mind. I remembered that. Let's go with what you have.

**Comment by bgchun** [comment]
@nhne This change breaks REEF on a YARN cluster. Did you test it on a YARN cluster?

**Comment by nhne** [comment]
@bgchun I didn't test on YARN cluster. I'll test it and let you know the result. Thanks.

**Comment by bgchun** [comment]
@nhne Please make sure to test changes on a YARN cluster for all the wake-related PRs.

**Comment by seojangho** [comment]
Hello, @nhne and @bgchun! I have investigated this issue.

When I modified `LoggingLinkListener` to write logs with `Level.SEVERE` on exception, I could hear it complaining like this:

```
java.lang.NoSuchMethodError: io.netty.buffer.ByteBuf.writeIntLE(I)Lio/netty/buffer/ByteBuf;
	at org.apache.reef.wake.remote.transport.netty.ChunkedReadWriteHandler.sizeAsByteArr(ChunkedReadWriteHandler.java:152)
	at org.apache.reef.wake.remote.transport.netty.ChunkedReadWriteHandler.write(ChunkedReadWriteHandler.java:128)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:658)
	at io.netty.channel.AbstractChannelHandlerContext.access$2000(AbstractChannelHandlerContext.java:32)
	at io.netty.channel.AbstractChannelHandlerContext$AbstractWriteTask.write(AbstractChannelHandlerContext.java:939)
	at io.netty.channel.AbstractChannelHandlerContext$WriteAndFlushTask.write(AbstractChannelHandlerContext.java:991)
	at io.netty.channel.AbstractChannelHandlerContext$AbstractWriteTask.run(AbstractChannelHandlerContext.java:924)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:380)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:357)
	at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:116)
	at java.lang.Thread.run(Thread.java:748)
```

So I added
```java
LOG.log(Level.INFO, "URL: {0}", this.getClass().getClassLoader().getResource("io/netty/buffer/ByteBuf.class"));
```
below [ChunkedReadWriteHandler.java#L149](https://github.com/apache/reef/blob/b8cecb612feaa96a86c031df0b6f6f7306f215b2/lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/ChunkedReadWriteHandler.java#L149)

and the result was
```
INFO: URL: jar:file:/home/jangho/.hadoop/share/hadoop/hdfs/lib/netty-all-4.0.23.Final.jar!/io/netty/buffer/ByteBuf.class
```

I think the project dependency needs some tweak.

**Comment by bgchun** [comment]
@seojangho Thanks for taking a look. Could you change REEF to use netty 4.0.23 and run YARN tests? 4.0.23 has the fix we want. Right?

**Comment by seojangho** [comment]
@bgchun Yes, it does. I'll post the result when it's get completed.

**Comment by seojangho** [comment]
@bgchun https://github.com/seojangho/reef/commit/1642187be3c10755f7e3730913b0188ae0e5dc35 passes all `runyarntests.sh` tests. ([Client log (stdout, stderr)](https://gist.github.com/seojangho/8aacd55c1454c5f7cd74ea6ae43a3954))

**Comment by bgchun** [comment]
@seojangho Thanks! Looks good.

**Comment by bgchun** [comment]
@seojangho checked the netty version of hadoop 2.7.x. The tar ball of hadoop 2.7.x comes with netty-all-4.0.23.Final.jar. It looks like it's best to downgrade to 4.0.23.Final to avoid the complicated dependency issue.
@seojangho Thanks for investigating this issue. This's very helpful. Can you send a PR that resolves https://issues.apache.org/jira/browse/REEF-1879?

**Comment by seojangho** [comment]
@bgchun Thanks for filing an issue. I'll send a PR soon.

**Comment by nhne** [comment]
@jangho Thank you for investigating this issue precisely! I really appreciate it.

@bgchun I'm so sorry that I missed that Netty can affect Yarn environment. I will check every tests more precisely next time.

**Pull request closed without merging**